### PR TITLE
Fix a bunch of dimensional and numerical issues with the Curved Poiss…

### PR DESCRIPTION
**Pull request by dsteinmo**
…on2D Sparse Matrix builder code.

* Was encountering stack smashing initially, followed by a bunch of inner dimension mismatches that needed clean-up.
* Also the way I was constructing the element-wise 'Broken LaPlacian' was down-right wrong.

### CODE REVIEW

File: src/Poisson2DSparseMatrix.cpp [hunk 1]
```diff
@@ -139,17 +145,75 @@ namespace blitzdg{
...
+
+                // Compute geometric factors on this element.
+                xr1 = blitz::sum(gDr(ii, jj)*xLocal(jj), jj);
+                xs1 = blitz::sum(gDs(ii, jj)*xLocal(jj), jj);
+
+                yr1 = blitz::sum(gDr(ii, jj)*yLocal(jj), jj);
+                ys1 = blitz::sum(gDs(ii, jj)*yLocal(jj), jj);
+
+                J1 = -xs1*yr1 + xr1*ys1;
+
```
Comment (k7miller) [inline src/Poisson2DSparseMatrix.cpp:176]: Any chance the `J1` can be zero here? Is that something we should check and issue an error if so?

File: src/Poisson2DSparseMatrix.cpp [hunk 2]
```diff
@@ -139,17 +145,75 @@ namespace blitzdg{
...
+                // Compute geometric factors on the neighboring element.
+                xr2 = blitz::sum(gDr(ii, jj)*xLocal2(jj), jj);
+                xs2 = blitz::sum(gDs(ii, jj)*xLocal2(jj), jj);
+
+                yr2 = blitz::sum(gDr(ii, jj)*yLocal2(jj), jj);
+                ys2 = blitz::sum(gDs(ii, jj)*yLocal2(jj), jj);
+
+                J2 = -xs1*yr1 + xr1*ys1;
+
```
Comment (k7miller) [inline src/Poisson2DSparseMatrix.cpp:210]: Same question as above for `J1`, can `J2` be zero and do we need to handle that case?

File: src/Poisson2DSparseMatrix.cpp [hunk 3]
```diff
@@ -139,17 +145,75 @@ namespace blitzdg{
...
+                rx2 = ys2/J2;
+                sx2 =-yr2/J2;
+                ry2 =-xs2/J2;
+                sy2 = xr2/J2;
+
+                real_matrix_type gDx2(NGauss, Np), gDy2(NGauss, Np);
+                gDx2 = 0.; gDy2 = 0.;
```
Comment (k7miller) [inline src/Poisson2DSparseMatrix.cpp]: Same as above.

File: src/Poisson2DSparseMatrix.cpp [hunk 4]
```diff
@@ -139,17 +145,75 @@ namespace blitzdg{
...
+                    for (index_type j=0; j < Np; ++j) {
+                        gDx2(i, j) = rx2(i) * gDr(i, j) + sx2(i) * gDs(i, j);
+                        gDy2(i, j) = ry2(i) * gDr(i, j) + sy2(i) * gDs(i, j);
+                    }
+                }
+
+                // Evaluate spatial derivatives of  Lagrange basis function at Gauss nodes
+                // on both faces, '-' and '+'.
                 real_vector_type gnx(NGauss), gny(NGauss), gw(NGauss);
                 gnx = 0.; gny = 0.; gw = 0.;
```
Comment (k7miller) [inline src/Poisson2DSparseMatrix.cpp]: Same as above.

File: src/Poisson2DSparseMatrix.cpp [hunk 5]
```diff
@@ -163,18 +227,17 @@ namespace blitzdg{
                 real_matrix_type gDnM(NGauss, Np), gDnP(NGauss, Np);
                 gDnM = 0.; gDnP = 0;
```
Comment (k7miller) [inline src/Poisson2DSparseMatrix.cpp]: Same as above. Note: `gDnP = 0.0`.

**Review by k7miller — APPROVED** [review]
Looks good to me Frank, just a few minor comments. Nice work.

File: src/Poisson2DSparseMatrix.cpp (hunk 1 above)
Comment (dsteinmo) [reply src/Poisson2DSparseMatrix.cpp:176]: If J1 is zero or changes sign it means it's somehow non-uniquely mappable to the standard element, which is bad. We probably do want to issue an error if any element is zero to within some tolerance.

File: src/Poisson2DSparseMatrix.cpp (hunk 2 above)
Comment (dsteinmo) [reply src/Poisson2DSparseMatrix.cpp:210]: Ditto explanation as J1.

File: src/Poisson2DSparseMatrix.cpp (hunk 1 above)
Comment (dsteinmo) [reply src/Poisson2DSparseMatrix.cpp:176]: This was apparently resolved in a separate commit sometime...

**Pull request merged**